### PR TITLE
perf(xidmap): Use btree with hash of keys for xidmap

### PR DESCRIPTION
**Pull request by ahsanbarkati**
Add b+ tree as an underlying data structure for xidmap.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6902)

 

Docs Preview: [Dgraph Preview](https://dgraph-a3c638ee78-109400.surge.sh)

### CODE REVIEW

**Comment by ahsanbarkati** [comment]
___
*[xidmap/xidmap.go, line 206 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6902#-MMLga1NCNvd-gzoGOcK:-MMLga1NCNvd-gzoGOcL:b-p4fxdx) ([raw file](https://github.com/dgraph-io/dgraph/blob/f348b73b27664918885f850afb07affc15a4f116/xidmap/xidmap.go#L206)):*
> ```Go
> 			x.Panic(err)
> 		}
> 	}
> ```

Push it to a channel.

**Comment by ahsanbarkati** [comment]
___
*[xidmap/xidmap.go, line 81 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6902#-MMLgfK400SyWCON-Tvz:-MMLgfK400SyWCON-Tw-:buaow0o) ([raw file](https://github.com/dgraph-io/dgraph/blob/f348b73b27664918885f850afb07affc15a4f116/xidmap/xidmap.go#L81)):*
> ```Go
> 	return h.Sum64()
> }
> ```

Use farm fingerprint

**Comment by ahsanbarkati** [comment]
___
*[xidmap/xidmap.go, line 206 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6902#-MMLga1NCNvd-gzoGOcK:-MMLgoJh4sith0FoRnWF:b-lay9nu) ([raw file](https://github.com/dgraph-io/dgraph/blob/f348b73b27664918885f850afb07affc15a4f116/xidmap/xidmap.go#L206)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

Push it to a channel.
</blockquote></details>

Batch the kvs to 64 and then push them to the channel.

File: xidmap/xidmap.go [hunk 1]
```diff
@@ -94,6 +92,14 @@ func New(zero *grpc.ClientConn, db *badger.DB) *XidMap {
 			tree: z.NewTree(100 << 20),
 		}
 	}
+
+	xm.kvChan = make(chan []kv, 64)
+
+	for i := 0; i < 16; i++ {
+		xm.wg.Add(1)
+		go xm.dbWriter()
```
Comment (jarifibrahim) [inline xidmap/xidmap.go:101]: The db could be nil. You should start these go routines inside the following if block

File: xidmap/xidmap.go [hunk 2]
```diff
@@ -94,6 +92,14 @@ func New(zero *grpc.ClientConn, db *badger.DB) *XidMap {
 			tree: z.NewTree(100 << 20),
 		}
 	}
+
+	xm.kvChan = make(chan []kv, 64)
```
Comment (jarifibrahim) [inline xidmap/xidmap.go:97]: Move this to where we declare the struct.

File: xidmap/xidmap.go [hunk 3]
```diff
@@ -163,15 +168,26 @@ func (m *XidMap) CheckUid(xid string) bool {
...
 	defer sh.Unlock()
-	sh.tree.Set(m.getHash(xid), uid)
+	sh.tree.Set(farm.Fingerprint64([]byte(xid)), uid)
+}
+
+func (m *XidMap) dbWriter() {
+	defer m.wg.Done()
+	for buf := range m.kvChan {
+		for _, kv := range buf {
+			if err := m.writer.Set(kv.key, kv.value); err != nil {
```
Comment (jarifibrahim) [inline xidmap/xidmap.go:187]: x.Panic(m.writer....)

File: xidmap/xidmap.go [hunk 4]
```diff
@@ -189,19 +205,24 @@ func (m *XidMap) AssignUid(xid string) (uint64, bool) {
...
 		binary.BigEndian.PutUint64(uidBuf[:], newUid)
-		if err := m.writer.Set([]byte(xid), uidBuf[:]); err != nil {
-			x.Panic(err)
+		m.kvBuf = append(m.kvBuf, kv{key: []byte(xid), value: uidBuf[:]})
+
+		if len(m.kvBuf) == 64 {
+			newBuf := make([]kv, len(m.kvBuf))
+			copy(newBuf, m.kvBuf)
+			m.kvChan <- newBuf
+			m.kvBuf = m.kvBuf[:0]
```
Comment (jarifibrahim) [inline xidmap/xidmap.go:226]:
```
m.kvChan <- m.kvBuf
m.kvBuf = make([]kv, 0, 64)
```

**Review by jarifibrahim — APPROVED** [review]
Run this locally as well.

**Pull request merged**